### DISCUSSION TITLE
Optimizing np.where and np.select

### DISCUSSION THREAD

**Discussion by asmeurer** [Quansight-Labs/metadsl#125] [Ideas]
This is motivated by https://github.com/sympy/sympy/issues/15014.

Take something like `np.select`, or if you prefer, nested `np.where` (`select` is not a ufunc).

`np.where(cond, x, y)` takes a boolean array `cond` and chooses `x` where it is True and `y` where it is False. 

The issue is if `x` and `y` are not arrays but expressions. For instance, `np.where(x <= 0, 0, np.log(x))`. This will evaluate `np.log` on every element of `x`, when it only should be evaluated on the positive elements. This is especially problematic if `x` is a non-NumPy scalar, but even if it is an array, it is wasteful. 

A more efficient way is to use a mask

```py
>>> x = np.array([-1, 0, 1])
>>> cond = x <= 0
>>> out = np.empty(x.shape)
>>> out[cond] = 0
>>> out[~cond] = np.log(x[~cond])
>>> out
array([0., 0., 0.])
```

If there is more than one variable, you need to use `np.broadcast`. For nested `np.where` (or `select`), you should chain the conditions (see the SymPy issue for an example).

**Comment by saulshanabrook** [comment]
This is the example from the  the [`numpy.select`](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.select.html)) docs:

```python
x = np.arange(10)
condlist = [x<3, x>5]
choicelist = [x, x**2]
np.select(condlist, choicelist)
```

So if we had a function like this:

```python
def t(x):
    condlist = [x<3, x>5]
    choicelist = [x, x**2]
    return np.select(condlist, choicelist)
```

Then the way I see us jitting this (assuming input array is one dimensional) would be:


```python
def t(x):
    new = np.empty(x.shape[0])
    for i in range(x.shape[0]):
        if x[i] < 3:
             new[i] = x[i]
        elif x[i] > 5:
             new[i] = x[i] **2
        else:
             new[i] = 0
     return new
```

I think that is similar to what you outlined, but done in a looping manner instead of as composite operations.

**Comment by asmeurer** [comment]
I agree that version would be better, if it ends up being optimized by numba or compiled to C, because the mask version requires looping over the condition array multiple times to get the mask. 

Does/will uarray have a notion of different backends, such as pure Python NumPy, jitted NumPy, C, etc., with different optimizations done depending on the backend?

**Comment by datnamer** [comment]
How are these loops generated? Is there an abstract "for i in a" that drops to a get index? 

What about backends that don't have fast elementwise access like GPU

**Comment by saulshanabrook** [comment]
> Does/will uarray have a notion of different backends, such as pure Python NumPy, jitted NumPy, C, etc., with different optimizations done depending on the backend?

Yep! But right now only AST of numpy expressions is implemented. That code is all seperate through and the core can be used with other backends.

> How are these loops generated? Is there an abstract "for i in a" that drops to a get index?

There is a concept of a "loop" over some range. [here](https://github.com/Quansight-Labs/uarray/blob/8eb476d83418d520c2d037de02b3f19fcb81bdd1/uarray/core/naturals.py#L51-L61) is the abstract definition.

[This](https://github.com/Quansight-Labs/uarray/blob/8eb476d83418d520c2d037de02b3f19fcb81bdd1/uarray/numpy/ast.py#L124-L149) is the logic for compiling that loop in the numpy AST context.

> What about backends that don't have fast elementwise access like GPU

I need to investigate this more, but the way we generate the looping code is pretty flexible. If we have some [number of threads in parallel](https://devblogs.nvidia.com/even-easier-introduction-cuda/#picking-up-the-threads), we should be able to divide the computed array between them, and have each do their part.

**Comment by datnamer** [comment]
Very interesting. Can you point me to an example of where and how this abstract loop construct is used in some complex logic (conditionals etc)? 

And is there subtype polymorphism of multiple levels or is it just abstract -> backend? Or there could be some category theory functional thing of which I'm not aware.
